tensorpipe/CMakeLists.txt

```diff
@@ -195,24 +207,41 @@ endif()
 tp_conditional_backend(
   TP_ENABLE_IBV "Enable InfiniBand transport" "LINUX")
 if(TP_ENABLE_IBV)
-  target_sources(tensorpipe PRIVATE
-    common/epoll_loop.cc
-    common/ibv.cc
-    transport/ibv/connection_impl.cc
-    transport/ibv/context_impl.cc
-    transport/ibv/error.cc
-    transport/ibv/factory.cc
-    transport/ibv/listener_impl.cc
-    transport/ibv/reactor.cc
-    transport/ibv/sockaddr.cc
-    transport/ibv/utility.cc)
-  set(TP_PUBLIC_HEADERS ${TP_PUBLIC_HEADERS}
-    transport/ibv/error.h
-    transport/ibv/factory.h
-    transport/ibv/utility.h)
+  list(APPEND TP_SRCS
+    "${CMAKE_CURRENT_SOURCE_DIR}/common/epoll_loop.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/common/ibv.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/connection_impl.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/context_impl.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/error.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/factory.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/listener_impl.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/reactor.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/sockaddr.cc"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/utility.cc"
+    )
+  list(APPEND TP_HDRS
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/error.h"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/factory.h"
+    "${CMAKE_CURRENT_SOURCE_DIR}/transport/ibv/utility.h"
+    )
   set(TENSORPIPE_HAS_IBV_TRANSPORT 1)
 endif()
 
+if(TP_USE_CUDA)
+  list(APPEND TP_SRCS ${TP_CUDA_SRCS})
+  list(APPEND TP_HDRS ${TP_CUDA_HDRS})
+endif()
+
+add_library(tensorpipe ${TP_STATIC_OR_SHARED} ${TP_SRCS})
+
+find_package(uv REQUIRED)
+target_link_libraries(tensorpipe PRIVATE uv::uv)
+
+if(TP_USE_CUDA)
+  find_package(CUDA REQUIRED)
+  target_link_libraries(tensorpipe PUBLIC ${CUDA_LIBRARIES})
+endif()
```

Comment:
Yes, `TP_USE_CUDA` is used few places. I tried to follow the structure/order 

- Get all the source/header files
- Add into `add_library()` or `add_executable()`
- Find and add deps packages to be linked
-  Add include directories and final the `install()` command.

I will break this PR and make the cmake file changes in a separate PR and try to include above comments into it.

Reply:
I have tried few of the changes in PR [391] (https://github.com/pytorch/tensorpipe/pull/391)

- Cuda headers are installed into the library, so keeping headers and sources in single list/variable is not feasible.
- Since the cmake command takes only few seconds to run, moving `find_package(CUDA required)` to start may not help so much.
- Can maintain a list `TP_DEPS` but this will lead to maintaining few more list to handle the different parameters to the `target_link_libraries()` calls for uv and CUDA and also take care of PUBLIC and PRIVATE. Please let me know your opinion, if OK I can add them. If cmake had maps it would have helped but it doesn't support it.